controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -9,515 +9,561 @@
+        synchronized (epochTxnLock) {
+            epochTxnMap.putIfAbsent(epoch, new HashSet<>());
+        }
         return CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public CompletableFuture<Void> scaleOldSegmentsSealed(List<Integer> sealedSegments, List<Integer> newSegments, long timestamp) {
-        Preconditions.checkNotNull(sealedSegments);
-        Preconditions.checkArgument(sealedSegments.size() > 0);
-
-        List<List<Integer>> predecessors = new ArrayList<>();
-        List<SimpleEntry<Double, Double>> keyRanges = newSegments.stream().map(x -> {
-            Segment segment = FutureHelpers.getAndHandleExceptions(getSegment(x), RuntimeException::new);
-            return new SimpleEntry<>(segment.getKeyStart(), segment.getKeyEnd());
-        }).collect(Collectors.toList());
-        for (int i = 0; i < keyRanges.size(); i++) {
-            predecessors.add(new ArrayList<>());
+    CompletableFuture<Void> deleteEpochNode(int epoch) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        synchronized (epochTxnLock) {
+            if (epochTxnMap.getOrDefault(epoch, Collections.emptySet()).isEmpty()) {
+                epochTxnMap.remove(epoch);
+                result.complete(null);
+            } else {
+                result.completeExceptionally(new DataExistsException(""));
+            }
         }
-
-        int start = segments.size();
-        // assign status, end times, and successors to sealed segments.
-        // assign predecessors to new segments
-        for (Integer sealed : sealedSegments) {
-            InMemorySegment segment = segments.get(sealed);
-            List<Integer> successors = new ArrayList<>();
-
-            for (int i = 0; i < keyRanges.size(); i++) {
-                if (segment.overlaps(keyRanges.get(i).getKey(), keyRanges.get(i).getValue())) {
-                    successors.add(newSegments.get(i));
-                    predecessors.get(i).add(sealed);
-                }
+        return result;
+    }
+
+    @Override
+    CompletableFuture<Integer> createNewTransaction(UUID txId, long timestamp, long leaseExpiryTime, long maxExecutionExpiryTime,
+                                                    long scaleGracePeriod) {
+        Preconditions.checkNotNull(txId);
+
+        final CompletableFuture<Integer> result = new CompletableFuture<>();
+        final Data<Integer> txnData = new Data<>(
+                new ActiveTxnRecord(timestamp, leaseExpiryTime, maxExecutionExpiryTime, scaleGracePeriod, TxnStatus.OPEN)
+                .toByteArray(), 0);
+        activeTxns.putIfAbsent(txId.toString(), txnData);
+        int epoch = activeEpoch.get();
+        synchronized (epochTxnLock) {
+            if (!epochTxnMap.containsKey(epoch)) {
+                result.completeExceptionally(new DataNotFoundException("epoch"));
+            } else {
+                epochTxnMap.computeIfPresent(epoch, (x, y) -> {
```

Comment:
we are using compute because we want to update the 'set' in the 'value'. otherwise we will have to do a get followed by a put. 
i have now replaced computeIfPresent with compute.